proxy/envoy/resources.go

```diff
@@ -252,34 +254,126 @@ func (route *HTTPRoute) CatchAll() bool {
 	return len(route.Headers) == 0 && route.Path == "" && route.Prefix == "/"
 }
 
-// CombinePathPrefix checks that the route applies for a given path and prefix
-// match and updates the path and the prefix in the route. If the route is
-// incompatible with the path or the prefix, returns nil.  Either path or
-// prefix must be set but not both.  The resulting route must match exactly the
-// requests that match both the original route and the supplied path and
-// prefix.
-func (route *HTTPRoute) CombinePathPrefix(path, prefix string) *HTTPRoute {
+const (
+	// MatchPath indicates if the route rule or match condition uses exact path match
+	MatchPath = iota
+	// MatchPrefix indicates if the route rule or match condition uses prefix path match
+	MatchPrefix = iota
+	// MatchRegex indicates if the route rule or match condition uses regex path match
+	MatchRegex = iota
+)
+
+// CombinePathPrefixRegex checks that the route applies for a given path, prefix
+// and regex match and updates the path, prefix, and regex in the route. If the
+// route is incompatible with these fields, returns nil. One of path or prefix
+// or regex must be set.  The resulting route must match exactly the requests
+// that match both the original route and the supplied path, prefix and regex.
+func (route *HTTPRoute) CombinePathPrefixRegex(path, prefix, regex string) *HTTPRoute {
+
+	var ingressMatch, routeMatch int
+
+	// assumption is that the validation engine has ensured that only path/prefix/regex is set
+	// for both ingress and route
 	switch {
-	case path == "" && route.Path == "" && strings.HasPrefix(route.Prefix, prefix):
-		// pick the longest prefix if both are prefix matches
-		return route
-	case path == "" && route.Path == "" && strings.HasPrefix(prefix, route.Prefix):
-		route.Prefix = prefix
-		return route
-	case prefix == "" && route.Prefix == "" && route.Path == path:
-		// pick only if path matches if both are path matches
-		return route
-	case path == "" && route.Prefix == "" && strings.HasPrefix(route.Path, prefix):
-		// if mixed, pick if route path satisfies the prefix
-		return route
-	case prefix == "" && route.Path == "" && strings.HasPrefix(path, route.Prefix):
-		// if mixed, pick if route prefix satisfies the path and change route to path
-		route.Path = path
-		route.Prefix = ""
-		return route
-	default:
+	case path != "":
+		ingressMatch = MatchPath
+	case prefix != "":
```

Comment:
Is it still the case that either path or prefix is set, even for regex?

Reply:
yes. Only one is allowed. path/prefix/regex

Reply:
I see, that's not quite obvious from here https://github.com/istio/pilot/blob/master/proxy/envoy/header.go#L28